Implement counting physical and logical lines

oneeyeman1 opened the pull request:
This PR will close ticket #12366.

It updates all 3 major ports (MSW, GTK and Cocoa) as well as the sample and documentation.

In the sample I show usage of old GetNumberOfLines() as well as new functionality.

oneeyeman1 committed:
Implement counting physical/logical lines in text control

oneeyeman1 committed:
Implement counting lines in GTK

oneeyeman1 committed:
Implement getting number of physical/logical lines in MSW

oneeyeman1 committed:
Revert "Implement getting number of physical/logical lines in MSW"

This reverts commit 28a8d987daffaaaf71529489db232d7d73e7b50b.

oneeyeman1 committed:
Implement physical/logical line count on MSW

oneeyeman1 committed:
Fixing the documentation

oneeyeman1 commented on include/wx/gtk/textctrl.h:
```diff
@@ -50,6 +50,8 @@ class WXDLLIMPEXP_CORE wxTextCtrl: public wxTextCtrlBase
     virtual wxString GetLineText(long lineNo) const wxOVERRIDE;
     virtual int GetNumberOfLines() const wxOVERRIDE;
+    virtual unsigned GetPhysicalLineCount() const wxOVERRIDE;
+    virtual unsigned GetLogicalLineCount() const wxOVERRIDE;
```
Code here: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/TextLayout/Tasks/CountLines.html.

As you can see OSX uses unsigned to have the line count.

catalinr replied:
What about gtk?

oneeyeman1 replied:
Well it uses guint and not just gint, right?
Also, I believe you should keep the type for virtual function...

catalinr replied:
> Well it uses guint and not just gint, right?

[gint
gtk_text_iter_get_line (const GtkTextIter *iter)](https://developer.gnome.org/gtk3/stable/GtkTextIter.html#gtk-text-iter-get-line)

oneeyeman1 replied:
@catalinr ,
And you missed the second portion of my comment...

catalinr replied:
And you are missing the point...
You chose unsigned int "because OSX uses it", when for a reason just as good you could have chosen int (i.e. "because GTK uses it") and at the same time could have been _consistent_.

oneeyeman1 replied:
@catalinr ,
From https://www.tutorialspoint.com/cplusplus/cpp_data_types.htm:

int | 4bytes | -2147483648 to 2147483647
unsigned int | 4bytes | 0 to 4294967295

Do you know a scenario under which the number of lines in the control will be a negative number?
I think the original author of the API was wrong.
I'd rather fix the original API, but I'm afraid this will break the user code.

vid512 replied:
Scenario with more than 2 million lines is just as unrealistic. But I can imagine a scenario, where some SearchLine function would return value -1 as 'not found', like many other wx functions do. And of course, you want to compare and do arithmetics which involves both line count and line index, so you need to have them in same type. Pretty much all other functions use signed value for line number. Should callers to cast their values around every use of these functions, not to get warnings? Changing the de-facto rule for just these two methods would introduce inconsistency in the interface, probably produce lots of signed/unsigned warnings, and bring no gain.

oneeyeman1 replied:
@vadz , so what should OSX version return? Also int? Like I said initially Cocoa API chooses unsigned int for some reason. I will check ifif OSX will compile without any warnings anyway.

vadz replied:
Of course all ports should return the same type. Warnings can be suppressed if really necessary.

oneeyeman1 commented on samples/text/text.cpp:
```diff
@@ -278,6 +278,21 @@ class MyFrame: public wxFrame
 
+    void OnGetPhysicalCount(wxCommandEvent &WXUNUSED(event))
+    {
+        wxMessageBox( wxString::Format( "Number of physical lines is %d", m_panel->m_textrich->GetPhysicalLineCount() ) );
```
I presume you are talking about spaces here.
I'm not blind but for me it is easier to see "(" with spaces, especially when there multiple of them.

catalinr replied:
Personal taste could have its place in new code, but in this file where previous messages (surprisingly) have the wxW preferred formatting, why try to change it? You also have some good examples for improving readability: split a long line on multiple lines.

oneeyeman1 commented on include/wx/textctrl.h:
```diff
@@ -555,7 +555,8 @@ class WXDLLIMPEXP_CORE wxTextAreaBase
     virtual int GetNumberOfLines() const = 0;
-
+    virtual unsigned GetPhysicalLineCount() const;
+    virtual unsigned GetLogicalLineCount() const;
```
That was my initial idea, but because then wxRTC and wxSTC will need the implementation, it was abandoned.
And I don't know their code well enough to submit them for those controls.

vid512 replied:
Placing declaration here (in wxTextAreaBase) means these two methods will be present in everything that derives from wxTextAreaBase or wxTextCtrlIface: That includes wxRichtTextCtrl, which would do better with more distinction (paragraph end vs. line break), and wxSearchCtrlBaseBaseClass which is never multiline.

Wouldn't it be better to move these declarations into wxTextCtrl() class? Then, they could be declared pure virtual and implementations would be needed only in wxTextCtrl control for every platform. Interface can be added to other controls, when someone feels like actually implementing it.

oneeyeman1 commented on src/gtk/textctrl.cpp:
```diff
@@ -1343,6 +1343,29 @@ int wxTextCtrl::GetNumberOfLines() const
+unsigned wxTextCtrl::GetPhysicalLineCount() const
+{
+    GtkTextIter end;
+    guint numberOfLines;
```
Sorry - old habit.
I did start as a C programmer and then moved to C++.

oneeyeman1 commented on src/gtk/textctrl.cpp:
```diff
@@ -1343,6 +1343,29 @@ int wxTextCtrl::GetNumberOfLines() const
+    guint numberOfLines;
+    GtkTextBuffer *buffer = gtk_text_view_get_buffer( GTK_TEXT_VIEW( m_text ) );
+    gtk_text_buffer_get_end_iter( buffer, &end );
+    numberOfLines = gtk_text_iter_get_line( &end ) + 1;
```
Vadim was talking about deprecating the GetNumberOfLines() function for both GTK and OSX. I wanted to create a native implementation for those 2 before this deprecation - that's all.

oneeyeman1 commented on interface/wx/textctrl.h:
```diff
@@ -1371,6 +1371,26 @@ class wxTextCtrl : public wxControl,
+    /*
+        Return number of logical lines in the text buffer.
+
+        This number corresponds to the number of soft carriage-return in
```
Do you have something in mind?
I just tried to explain the difference between them. And I didn't even tried to explain the difference for the original function as it is platform specific.

catalinr replied:
The old function refers to `newline characters`. You just introduced "hard carriage-return" and "soft carriage-return", really?

oneeyeman1 replied:
I think that's how it is referenced in documentation.
But I guess I should rewrite it with the `newline character` phrase.

oneeyeman1 replied:
@catalinr ,
I tried to rewrite it using `newline character`.
Do you think it is better?

oneeyeman1 commented on samples/text/text.cpp:
```diff
@@ -278,6 +278,21 @@ class MyFrame: public wxFrame
+
+    void OnGetLineCount(wxCommandEvent &WXUNUSED(event))
+    {
+        wxMessageBox( wxString::Format( "Number of lines is %d", m_panel->m_textrich->GetNumberOfLines() ) );
```
For the time being I just wanted to show it as a difference between 2 new ones and 1 old.
Besides Vadim were talking about deprecating it for GTK/OSX only.

If he says "remove it" - I'll do it.

oneeyeman1 committed:
Trying to improve documentation on the new functions

oneeyeman1 committed:
Improve readability of the sample

oneeyeman1 committed:
Implement PR suggestion

oneeyeman1 committed:
More PR suggestions

oneeyeman1 commented on src/osx/textctrl_osx.cpp:
```diff
@@ -716,6 +726,16 @@ int wxTextWidgetImpl::GetNumberOfLines() const
 
+unsigned wxTextWidgetImpl::GetPhysicalLineCount() const
+{
+    return 0;
```
@catalinr ,
I don't know anything about how OSX implements stuff.
But I presume that textctrl_osx.cpp is/was just a generic interface for both Carbon and Cocoa.

Now I did implement that functionality in Cocoa, but since I was not sure what this file is I just left it alone.

vadz replied:
This should be pure virtual or return -1 too. Returning 0 is definitely wrong.

oneeyeman1 committed:
Implement PR suggestions on Cocoa

oneeyeman1 committed:
Fix the definition of the function in MSW header

oneeyeman1 commented on src/osx/cocoa/textctrl.mm:
```diff
@@ -1258,6 +1258,27 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
+{
+    NSString *string = [m_textView string];
+    unsigned numberOfLines, index, stringLength = [string length];
+    for (index = 0, numberOfLines = 0; index < stringLength; numberOfLines++)
```
@catalinr ,
I just tried it.
Unfortunately it didn't work - the function returned "1" instead. So I left the declaration alone.
I did however fixed formatting and added the braces for the loop.

Tested on OSX 10.13.

oneeyeman1 commented:
@catalinr ,
Is there anything left here to do?
Did you try the PR to check whether it works as advertised?

I think it is time to as Stefan to chime in and re-check Cocoa piece...

Thank you.

catalinr commented:
> Is there anything left here to do?

Well, since you asked, these are some quick thoughts (and they probably do not differ that much from the previous ones):
- the new functions should return `int` not `unsigned`;
- I can't find any `hard newline` or `soft newline` in current docs, I find the syntagms far from self-explanatory, and even more so when reading only one of them, and finally, I don't understand what was wrong with existing `logical lines` / `physical lines` that are also clearly exemplified to eliminate any confusion;
- _still_ having formatting issues, even with the line that the suggestion was given for, and numerous talks about it, is IMHO a very bad sign (hint: look for `wxMessageBox( `);
- using `GetNumberOfLines()` in the sample with such a useless message is just as useless;
- one of the new implementations for GTK should use or be used in `GetNumberOfLines()` (oh, and should I mention formatting here too?);
- under OSX, at least `wxTextWidgetImpl::GetPhysicalLineCount()` should return `wxTextWidgetImpl::GetNumberOfLines()`. Why do you insist on it returning 0?

oneeyeman1 commented:
@catalinr ,
> 
> > Is there anything left here to do?
> 
> Well, since you asked, these are some quick thoughts (and they probably do not differ that much from the previous ones):

Yes, thank you for commenting.

> 
>     * the new functions should return `int` not `unsigned`;

I tried to explain the reason behind it.
I doubt there will be a scenario where the number of lines will be -5, which is perfectly legal for int, but not for unsigned. To make it even better - the return value should have been "unsigned long", to make the range of possible values even bigger, but that would probably be overkill. I doubt people will write something like "War and Piece" in wxWidgets-based program.
But this is the place where we can make a new API more suitable for what functionality can do - don't you think?

> 
>     * I can't find any `hard newline` or `soft newline` in current docs, I find the syntagms far from self-explanatory, and even more so when reading only one of them, and finally, I don't understand what was wrong with existing `logical lines` / `physical lines` that are also clearly exemplified to eliminate any confusion;

I think I saw somewhere a reference to "hard carriage return"/"soft carriage return", but it was a long time ago. And maybe now the computer literature moves on.
But I see that even Vadim refer to `physical lines` as `lines visible on the screen`. And I don't know what is good and bad here.
All I can say is that I'm an ESL person and judging by your name - you are as well. We can argue all we want, but never get a consensus.
But if you say that saying `logical lines`/`physical lines` is better choice - that is OK. I guess it can always be changed later.

I just tried to improve the documentation by completely removing "hard/soft" words there. Let me know if this is not enough.

> 
>     * _still_ having formatting issues, even with the line that the suggestion was given for, and numerous talks about it, is IMHO a very bad sign (hint: look for `wxMessageBox( `);

I tried to fix it in the sample.
Is there anything else? I did recheck and there is no extra spaces anywhere...

> 
>     * using `GetNumberOfLines()` in the sample with such a useless message is just as useless;

I tried to use the same wording as in the documentation.

> 
>     * one of the new implementations for GTK should use or be used in `GetNumberOfLines()` (oh, and should I mention formatting here too?);

I tried to explain this as well - Vadim wanted to deprecate this for GTK/OSX and use the new API.
Did you read my reply above?

> 
>     * under OSX, at least `wxTextWidgetImpl::GetPhysicalLineCount()` should return `wxTextWidgetImpl::GetNumberOfLines()`. Why do you insist on it returning 0?

One more time - it is possible that old API for GTK/OSX will go away.
So when this happens this function will need to be rewritten.

Now unless Vadim, chime in and explicitly ask me to deprecate the old API for those 2 (or maybe even all 3) platforms - I prefer to leave this functionality alone.

You can search for that discussion in the ML archive.

oneeyeman1 committed:
Implement PR suggestions

vadz reviewed (changes requested):
Thanks for your work on this and sorry for the delay with reviewing it, but now that I did it, there are 2 big problems with this PR.

First one is that I think it exchanges "physical" and "logical" meanings. "Logical" to me means corresponding to the number of hard line breaks in the string, i.e. doesn't change depending on the size of the text control or its style. "Physical" means the lines the user actually, physically, sees. I'm surprised you've chosen the opposite meanings for these functions, but even more surprised that neither @catalinr nor @vid512 who have reviewed the PR have objected to it. Maybe this means that things are less obvious than I thought. Perhaps we should use something less ambiguous, e.g. `GetNumberOf{Hard,Soft}Lines()` for symmetry with hard/soft line breaks?

Second problem is that the PR doesn't include any tests, so I have absolutely no idea whether the code actually works or not. Luckily, there is now a test in https://trac.wxwidgets.org/ticket/12366#comment:23 (thanks a lot @vid512!), but it needs to be integrated into this PR to actually run it for all platforms.

There are also other considerations, such as the complexity of these functions. This may be important for controls containing a lot of text, so if they're slow (and I think GTK versions might well be), we need to at least document this, so that the user code could consider caching the results if necessary.

But the most important things to do are to use consistent names for the functions and to test them.

vid512 commented:
About the names. GTK uses names "display line" for wrapped lines (https://developer.gnome.org/gtk3/stable/GtkTextView.html#gtk-text-view-forward-display-line). That's completely unambigous, and IMO a very good name. Question is, how to call the other one then.

Line in source text could be called "Paragraph" within context of wxTextCtrl, eg. GetParagraphCount. But this would probably be confusing be in wxRichTextCtrl, because wxRichTextCtrl supports line breaks within its paragraph. And wxRichText inherits from wxTextCtrl.

GTK documentation also kind-of uses name "text buffer Y coordinate" to clearly specify line number in original source text buffer, as opposed to display line. It's because GTK stores the text itself (not its displayed layout) in separate GtkTextBuffer object. Name "GetBufferLineCount" would be unambigous for our purpose, but probably would be bit confusing when viewed from different context (what buffer?).

I'd strongly suggest to use "GetDisplayLineCount" for what vadz calls "physical". As for the other, I see no clear winner. GetBufferLineCount, GetTextLineCount, GetLogicalLineCount all seem tolerable, but not great... 

Another possibility is simply GetWrappedLineCount vs. GetUnwrappedLineCount. And whoever deals with wxTE_NOWRAP can just use old GetLineCount (I mean, GetNumberOfLines) anyway.

vadz commented:
> About the names. GTK uses names "display line" for wrapped lines (https://developer.gnome.org/gtk3/stable/GtkTextView.html#gtk-text-view-forward-display-line). That's completely unambigous, and IMO a very good name. 

Thanks, this is indeed a good name and it's nice to reuse something familiar for the programmers under at least one of the 3 major platforms. I'm perfectly fine with using it.

> Question is, how to call the other one then.

It's less important if the "other" name is clear as it could be defined as just being the "other one".

> GTK documentation also kind-of uses name "text buffer Y coordinate" to clearly specify line number in original source text buffer, as opposed to display line. It's because GTK stores the text itself (not its displayed layout) in separate GtkTextBuffer object. Name "GetBufferLineCount" would be unambigous for our purpose, but probably would be bit confusing when viewed from different context (what buffer?).

Yes, exactly. Maybe we could use `GetValueLineCount()` to make it clear that it just counts the number of lines in `GetValue()`. In fact, I'm rather surprised by the fact that the corresponding function is (almost) pure virtual  in this PR, I would think that it would be implemented in common count as it's done for wxMSW, i.e. using `std::count()`. I have no idea if the PR uses more code for performance reasons or if there is some other explanation.

> I'd strongly suggest to use "GetDisplayLineCount" for what vadz calls "physical". As for the other, I see no clear winner. GetBufferLineCount, GetTextLineCount, GetLogicalLineCount all seem tolerable, but not great...

The two latter alternatives would be fine too.

> Another possibility is simply GetWrappedLineCount vs. GetUnwrappedLineCount.

I could also live with this.

oneeyeman1 commented:
@vadz , I have a question.
Are we, as you wrote in the ticket, deprecating GetNumberOfLines()? Because if we are I can include it in this PR...

Thx.

vadz commented:
> Are we, as you wrote in the ticket, deprecating GetNumberOfLines()?

No, I don't think we're going to deprecate it formally, there is no good reason to break/force to change all the existing code using it if it doesn't use wrapping at all. We should document that it doesn't work consistently between different platforms and point to the new methods in the documentation of the wrapping styles however.

oneeyeman1 commented:
This might help with #18809.

oneeyeman1 committed:
Rebase the branch to the latest master

oneeyeman1 committed:
Update Scintilla submodule

oneeyeman1 committed:
Fix trailing spaces

oneeyeman1 committed:
Use 'int' as a return type for a new function

oneeyeman1 committed:
Implement PR suggestion

oneeyeman1 committed:
Adding the test for the new functions

oneeyeman1 commented:
I changed the return type of the nre functions and added unit test of them.

I can also mark the old function as deprecated as Vadim suggested in the ticket.

Hopefully the test will pass and everything will be in order and ready for submission.

Thx.

oneeyeman1 committed:
Empty control return 1 line for consistency

oneeyeman1 commented:
@vid512 
> About the names. GTK uses names "display line" for wrapped lines (https://developer.gnome.org/gtk3/stable/GtkTextView.html#gtk-text-view-forward-display-line). That's completely unambigous, and IMO a very good name. Question is, how to call the other one then.

What is considered "display line"?
Is it the line currently displayed on the screen? Or is it the line that is in the buffer and wrapped? Or maybe it is partially displayed?

For GTK it kind of make sense because it use text buffer object.

Maybe call the functin GetWrappedLineCount() then? That would not be ambiguous at any level.

> 
> Line in source text could be called "Paragraph" within context of wxTextCtrl, eg. GetParagraphCount. But this would probably be confusing be in wxRichTextCtrl, because wxRichTextCtrl supports line breaks within its paragraph. And wxRichText inherits from wxTextCtrl.
> 
> GTK documentation also kind-of uses name "text buffer Y coordinate" to clearly specify line number in original source text buffer, as opposed to display line. It's because GTK stores the text itself (not its displayed layout) in separate GtkTextBuffer object. Name "GetBufferLineCount" would be unambigous for our purpose, but probably would be bit confusing when viewed from different context (what buffer?).
> 
> I'd strongly suggest to use "GetDisplayLineCount" for what vadz calls "physical". As for the other, I see no clear winner. GetBufferLineCount, GetTextLineCount, GetLogicalLineCount all seem tolerable, but not great...
> 
> Another possibility is simply GetWrappedLineCount vs. GetUnwrappedLineCount. And whoever deals with wxTE_NOWRAP can just use old GetLineCount (I mean, GetNumberOfLines) anyway.

I don't know.
For me "physical" line count corresponds to the number of '\n\r' characters as its physical characters and it just physically breaks the line.
And "logical" corresponds to the wrapping as its not user decision but logical breaks of the line.

Maybe it just me...

Thank you.

vadz commented:
If we use "display line" for wrapped/physical (because they're actually _physically_ shown on screen) lines, then we should use "text line" for unwrapped/logical (because this is what the _logic_ of the program considers to be a line).

oneeyeman1 commented:
I have no idea why GTK fails.

Applying following:

```
diff --git a/samples/minimal/minimal.cpp b/samples/minimal/minimal.cpp
index 5f32257c6b..0a610a0a08 100644
--- a/samples/minimal/minimal.cpp
+++ b/samples/minimal/minimal.cpp
@@ -67,6 +67,7 @@ public:
 private:
     // any class wishing to process wxWidgets events must use this macro
     wxDECLARE_EVENT_TABLE();
+    wxTextCtrl *m_text;
 };
 
 // ----------------------------------------------------------------------------
@@ -175,6 +176,9 @@ MyFrame::MyFrame(const wxString& title)
     CreateStatusBar(2);
     SetStatusText("Welcome to wxWidgets!");
 #endif // wxUSE_STATUSBAR
+    m_text = new wxTextCtrl( this, wxID_ANY, "", wxDefaultPosition, wxSize(400, 200 ), wxTE_MULTILINE );
+    m_text->SetValue("line1\nline2\nlong long line 3");
+    m_text->AppendText( "\nThis is a very long line to test logical line count in the text control." );
 }
 
 
@@ -188,6 +192,13 @@ void MyFrame::OnQuit(wxCommandEvent& WXUNUSED(event))
 
 void MyFrame::OnAbout(wxCommandEvent& WXUNUSED(event))
 {
+    auto physical = m_text->GetPhysicalLineCount();
+    auto logical = m_text->GetLogicalLineCount();
+    wxMessageBox( wxString::Format( "Physical line count is %d, logical line count is %d", physical, logical ) );
+    m_text->AppendText("\n\nMore text on line 5");
+    physical = m_text->GetPhysicalLineCount();
+    logical = m_text->GetLogicalLineCount();
+    wxMessageBox( wxString::Format( "Physical line count is %d, logical line count is %d", physical, logical ) );
     wxMessageBox(wxString::Format
                  (
                     "Welcome to %s!\n"
lines 1-36/36 (END)

```

yields correct results.

Running ob GTK3.
